File: jetbrains-plugin/src/main/kotlin/com/idebridge/jetbrains/symbol/SymbolRelocation.kt
package com.idebridge.jetbrains.symbol

import com.idebridge.jetbrains.protocol.Range
import com.idebridge.jetbrains.protocol.SymbolLocator

/**
 * Controlled relocation of a stale symbol reference (ADR-0003, amended by ADR-0018).
 *
 * A handle is a fast path bound to a session, an epoch, and a revision. When it no longer resolves,
 * the locator is the durable identity. Matching uses the locator's semantic fields — name, kind, and
 * container name when both sides declare one — inside its own document, with the selection range as
 * a tie-breaker only. Matching the fingerprint instead would fail for any symbol shifted by a single
 * line, which is exactly when a handle goes stale.
 *
 * This rule also exists in the VS Code adapter. Both are checked against
 * `packages/protocol/fixtures/vectors/symbol-relocation-vectors.json` so the same protocol cannot
 * answer differently depending on the IDE (ADR-0025).
 *
 * Nothing is ever guessed: no match fails closed, and several indistinguishable matches report every
 * candidate.
 */
public object SymbolRelocation {
    /** Upper bound on candidates reported with `AMBIGUOUS_SYMBOL`. */
    private const val MAX_CANDIDATES = 32

    /** A mapped symbol: its locator plus its declaration range and children. */
    public data class Draft(
        val locator: SymbolLocator,
        val range: Range,
        val children: List<Draft> = emptyList(),
    )

    public sealed interface Outcome {
        public data class Resolved(val draft: Draft) : Outcome

        public data object NotFound : Outcome

        public data class Ambiguous(val candidates: List<SymbolLocator>) : Outcome
    }

    public fun relocate(target: SymbolLocator, current: List<Draft>): Outcome {
        val matches = collectMatches(target, current)
        if (matches.isEmpty()) return Outcome.NotFound
        matches.singleOrNull()?.let { return Outcome.Resolved(it) }

        // Same name, kind, and container: the original selection range is the only discriminator
        // left, and it decides only when exactly one candidate still carries it.
        val exact = matches.filter { sameRange(it.locator.selectionRange, target.selectionRange) }
        exact.singleOrNull()?.let { return Outcome.Resolved(it) }
        return Outcome.Ambiguous(matches.take(MAX_CANDIDATES).map { it.locator })
    }

    private fun collectMatches(target: SymbolLocator, current: List<Draft>): List<Draft> {
        val matches = mutableListOf<Draft>()
        val pending = ArrayDeque(current)
        while (pending.isNotEmpty()) {
            val draft = pending.removeFirst()
            if (isSameIdentity(draft.locator, target)) matches.add(draft)
            pending.addAll(draft.children)
        }
        return matches
    }

    /**
     * Container name is compared as optional on both sides: a locator minted from a flat search
     * result legitimately lacks the container a hierarchical provider reports, and treating that
     * absence as a mismatch would make such a symbol unrelocatable.
     */
    private fun isSameIdentity(candidate: SymbolLocator, target: SymbolLocator): Boolean {
        if (candidate.documentUri != target.documentUri) return false
        if (candidate.name != target.name) return false
        // The declaration's syntactic form discriminates when both sides have it; `kind` does
        // the job only for adapters whose IDE supplies one, and is `unknown` for the rest.
        // Comparing `unknown` to `unknown` would make a field and a method indistinguishable.
        val candidateType = candidate.declarationType
        val targetType = target.declarationType
        if (candidateType != null && targetType != null) {
            if (candidateType != targetType) return false
        } else if (candidate.kind != target.kind) {
            return false
        }
        val candidateContainer = candidate.containerName
        val targetContainer = target.containerName
        if (candidateContainer == null || targetContainer == null) return true
        return candidateContainer == targetContainer
    }

    private fun sameRange(left: Range, right: Range): Boolean =
        left.start.line == right.start.line &&
            left.start.character == right.start.character &&
            left.end.line == right.end.line &&
            left.end.character == right.end.character
}

File: packages/vscode-extension/src/symbol-relocation.ts
/**
 * Controlled relocation of stale symbol references (ADR-0003, amended by ADR-0018).
 *
 * A handle is a fast path bound to a session, an epoch, and — for document symbols — a revision.
 * When it no longer resolves, the locator is the durable identity. Relocation matches on the
 * locator's semantic fields (name, kind, containerName) inside its own document and uses the
 * selection range only to break ties. Matching on the locator fingerprint instead would fail for
 * any symbol shifted by a single line, which is precisely when a handle goes stale.
 *
 * Relocation never guesses: zero matches fail closed with `STALE_SYMBOL`, and several
 * indistinguishable matches fail closed with `AMBIGUOUS_SYMBOL` carrying every candidate.
 */

import type { Range, SymbolLocator } from "@ide-bridge/protocol";

import type { SymbolDraft } from "./symbol-mapper.js";

/** Mirrors the `AMBIGUOUS_SYMBOL` schema invariant: ambiguity means at least two candidates. */
export type AmbiguousCandidates = [SymbolLocator, SymbolLocator, ...SymbolLocator[]];

export type RelocationOutcome =
  | { kind: "resolved"; draft: SymbolDraft }
  | { kind: "not-found" }
  | { kind: "ambiguous"; candidates: AmbiguousCandidates };

/** Upper bound on candidates reported with `AMBIGUOUS_SYMBOL`. */
const MAX_AMBIGUOUS_CANDIDATES = 32;

export function relocateSymbol(
  locator: SymbolLocator,
  drafts: readonly SymbolDraft[],
): RelocationOutcome {
  const matches = collectMatches(locator, drafts);
  if (matches.length === 0) return { kind: "not-found" };
  if (matches.length === 1) {
    const only = matches[0];
    if (only === undefined) return { kind: "not-found" };
    return { kind: "resolved", draft: only };
  }

  // Same name, kind, and container: the original selection range is the only remaining
  // discriminator, and it is authoritative only when exactly one candidate still carries it.
  const exact = matches.filter((draft) =>
    sameRange(draft.locator.selectionRange, locator.selectionRange),
  );
  const single = exact.length === 1 ? exact[0] : undefined;
  if (single !== undefined) return { kind: "resolved", draft: single };

  const [first, second, ...rest] = matches
    .slice(0, MAX_AMBIGUOUS_CANDIDATES)
    .map((draft) => structuredClone(draft.locator));
  if (first === undefined || second === undefined) return { kind: "not-found" };
  return { kind: "ambiguous", candidates: [first, second, ...rest] };
}

function collectMatches(locator: SymbolLocator, drafts: readonly SymbolDraft[]): SymbolDraft[] {
  const matches: SymbolDraft[] = [];
  const pending = [...drafts];
  while (pending.length > 0) {
    const draft = pending.pop();
    if (draft === undefined) continue;
    if (isSameSymbolIdentity(draft.locator, locator)) matches.push(draft);
    pending.push(...draft.children);
  }
  return matches;
}

/**
 * Container name is compared as an optional field: a locator minted from a flat search result may
 * legitimately lack the container that the hierarchical document provider reports, so an absent
 * container on either side is treated as "unspecified" rather than as a mismatch.
 */
function isSameSymbolIdentity(candidate: SymbolLocator, target: SymbolLocator): boolean {
  if (candidate.documentUri !== target.documentUri) return false;
  if (candidate.name !== target.name || candidate.kind !== target.kind) return false;
  if (candidate.containerName === undefined || target.containerName === undefined) return true;
  return candidate.containerName === target.containerName;
}

function sameRange(left: Range, right: Range): boolean {
  return (
    left.start.line === right.start.line &&
    left.start.character === right.start.character &&
    left.end.line === right.end.line &&
    left.end.character === right.end.character
  );
}

/** Finds the innermost symbol whose declaration range contains `position`. */
export function findSymbolAtPosition(
  drafts: readonly SymbolDraft[],
  position: { line: number; character: number },
): SymbolDraft | undefined {
  let found: SymbolDraft | undefined;
  for (const draft of drafts) {
    if (!rangeContains(draft.range, position)) continue;
    found = findSymbolAtPosition(draft.children, position) ?? draft;
    break;
  }
  return found;
}

function rangeContains(range: Range, position: { line: number; character: number }): boolean {
  return comparePosition(range.start, position) <= 0 && comparePosition(position, range.end) <= 0;
}

function comparePosition(
  left: { line: number; character: number },
  right: { line: number; character: number },
): number {
  return left.line === right.line ? left.character - right.character : left.line - right.line;
}
